pkg/protocol/protocol.go

```diff
@@ -101,6 +109,21 @@ func (p *Protocol) Run() {
 	p.TipManager = p.optsTipManagerProvider(p.mainEngine)
 	p.Events.TipManager.LinkTo(p.TipManager.Events())
 
+	// TODO: the BlockIssuer needs to be invalidated together with the TipManager when switching engines.
+	p.BlockIssuer = blockissuer.New(
+		lo.PanicOnErr(p.localPeer.Database().LocalPrivateKey()),
+		p.API,
+		p.mainEngine.IsBootstrapped,
+		func() (*iotago.Commitment, iotago.SlotIndex) {
+			return p.mainEngine.Storage.Settings().LatestCommitment(), p.mainEngine.Storage.Settings().LatestFinalizedSlot()
+		},
+		p.ProcessBlock,
+		p.TipManager,
+		p.optsBlockIssuerOptions...,
+	)
```

Comment:
I still think it would be better to also have providers for the protocol and its submodules